### PR TITLE
 Smart contracts tests - erc20 (C and SOL), bad / malicious SC

### PR DESCRIPTION
**Pull request by andreibancioiu**

### CODE REVIEW

File: integrationTests/vm/arwen/sctests/badC/bad.export [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+divideByZero
+badGetOwner1
+badGetBlockHash1
+badGetBlockHash2
+badGetBlockHash3
+badWriteLog1
+badWriteLog2
+badWriteLog3
+badWriteLog4
+badBigIntStorageStore1
```
Comment (sasurobert) [inline integrationTests/vm/arwen/sctests/badC/bad.export]: add an empty line

File: integrationTests/vm/arwen/sctests/corrupt/corrupt.wasm [hunk 2]
```diff
@@ -0,0 +1 @@
+this-is-a-corrupt-wasm
```
Comment (sasurobert) [inline integrationTests/vm/arwen/sctests/corrupt/corrupt.wasm]: empty line at endoffile

File: integrationTests/vm/arwen/sctests/emptyC/empty.c [hunk 3]
```diff
@@ -0,0 +1,2 @@
+void _main(void) {
+}
```
Comment (sasurobert) [inline integrationTests/vm/arwen/sctests/emptyC/empty.c]: empty line at endoffile

File: integrationTests/vm/arwen/sctests/erc20c/wrc20_arwen.c [hunk 4]
```diff
@@ -0,0 +1,311 @@
+#include "elrond/context.h"
```
Comment (sasurobert) [inline integrationTests/vm/arwen/sctests/erc20c/wrc20_arwen.c]: move all source files into a sources folder. all go files directly under arwen.

File: integrationTests/vm/arwen/erc20_test.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	assert.Equal(t, uint64(100), context.querySCInt("balanceOf(address)", [][]byte{context.Carol.Address}))
+}
+
+func Test_SOL_003(t *testing.T) {
+	context := setupTestContext(t)
+
+	context.deploySC("./testdata/erc20sol003/0-0-3.wasm", "")
+
+	// Minting
+	context.executeSC(&context.Owner, "transfer(address,uint256)@"+context.Alice.AddressHex()+"@"+formatHexNumber(1000))
```
Comment (miiu96) [inline integrationTests/vm/arwen/erc20_test.go]: maybe you can use some constants

File: integrationTests/vm/arwen/erc20_test.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	context := setupTestContext(t)
+
+	context.deploySC("./testdata/erc20c/wrc20_arwen.wasm", formatHexNumber(42000))
+
+	// Assertion
+	assert.Equal(t, uint64(42000), context.querySCInt("totalSupply", [][]byte{}))
+	assert.Equal(t, uint64(42000), context.querySCInt("balanceOf", [][]byte{context.Owner.Address}))
+
+	// Minting
+	context.executeSC(&context.Owner, "transferToken@"+context.Alice.AddressHex()+"@"+formatHexNumber(1000))
```
Comment (miiu96) [inline integrationTests/vm/arwen/erc20_test.go]: use some constants

File: integrationTests/vm/arwen/testdata/badC/bad.c [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+void badBigIntStorageStore1()
+{
+  bigInt number = bigIntNew(100);
+  bigIntStorageStore("test", number + 42);
+}
+
+void _main(void)
+{
+}
```
Comment (miiu96) [inline integrationTests/vm/arwen/testdata/badC/bad.c]: add an empty line at the end of the file

File: integrationTests/vm/arwen/testdata/erc20c/elrond/debug.h [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+int strlen(const byte *str)
+{
+    const byte *s;
+    for (s = str; *s; ++s);
+    return (s - str);
+}
+
+void myDebugPrintString(byte *data) {
+    //debugPrintString(data, strlen(data));
+}
```
Comment (miiu96) [inline integrationTests/vm/arwen/testdata/erc20c/elrond/debug.h]: add new line at the end of the file

File: integrationTests/vm/arwen/utils.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+	context.Accounts = vm.CreateInMemoryShardAccountsDB()
+	context.createAccount(&context.Owner)
+	context.createAccount(&context.Alice)
+	context.createAccount(&context.Bob)
+	context.createAccount(&context.Carol)
+}
+
+func (context *testContext) createAccount(participant *testParticipant) {
+	_ = vm.CreateAccount(context.Accounts, participant.Address, participant.Nonce, participant.Balance)
```
Comment (miiu96) [inline integrationTests/vm/arwen/utils.go]: maybe you can put an assert here to check if an error appears

File: integrationTests/vm/arwen/utils.go [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+		Data:     txData,
+	}
+
+	err := context.TxProcessor.ProcessTransaction(tx, context.Round)
+	if err != nil {
+		assert.FailNow(context.T, err.Error())
+	}
+
+	owner.Nonce++
+	_, _ = context.Accounts.Commit()
```
Comment (miiu96) [inline integrationTests/vm/arwen/utils.go]: maybe you can put an assert here to check if an error appears

File: integrationTests/vm/arwen/utils.go [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+func getSCCode(fileName string) string {
+	code, _ := ioutil.ReadFile(fileName)
+	codeEncoded := hex.EncodeToString(code)
+	return codeEncoded
```
Comment (miiu96) [inline integrationTests/vm/arwen/utils.go]: add empty line

File: integrationTests/vm/arwen/utils.go [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+	}
+
+	sender.Nonce++
+	_, _ = context.Accounts.Commit()
+}
+
+func (context *testContext) querySCInt(function string, args [][]byte) uint64 {
+	bytes := context.querySC(function, args)
+	result := big.NewInt(0).SetBytes(bytes).Uint64()
+	return result
```
Comment (miiu96) [inline integrationTests/vm/arwen/utils.go]: add empty line

File: integrationTests/vm/arwen/utils.go [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+	}
+
+	firstResult := vmOutput.ReturnData[0]
+	return firstResult
+}
+
+func formatHexNumber(number uint64) string {
+	bytes := big.NewInt(0).SetUint64(number).Bytes()
+	str := hex.EncodeToString(bytes)
+	return str
```
Comment (miiu96) [inline integrationTests/vm/arwen/utils.go]: add empty line

File: integrationTests/vm/arwen/utils.go [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+	context.TxProcessor = vm.CreateTxProcessorWithOneSCExecutorWithVMs(context.Accounts, vmContainer, blockChainHook)
+	context.ScAddress, _ = blockChainHook.NewAddress(context.Owner.Address, context.Owner.Nonce, factory.ArwenVirtualMachine)
+	context.QueryService, _ = smartContract.NewSCQueryService(vmContainer, math.MaxInt32)
+	return context
+}
+
+func (context *testContext) initAccounts() {
+	context.Owner = testParticipant{}
+	context.Owner.Address = []byte{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o', 'o'}
+	context.Owner.Nonce = uint64(1)
```
Comment (miiu96) [inline integrationTests/vm/arwen/utils.go]: maybe you can create a variable initialNonce := uint64(1)

File: integrationTests/vm/arwen/utils.go (hunk 9 above)
Comment (andreibancioiu) [reply integrationTests/vm/arwen/utils.go]: Added a check.

File: integrationTests/vm/arwen/testdata/erc20c/elrond/bigInt.h [hunk 15]
```diff
@@ -0,0 +1,28 @@
+#include "types.h"
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply integrationTests/vm/arwen/testdata/erc20c/elrond/bigInt.h]: We currently do this for inclusion of `byte` typedef: https://github.com/ElrondNetwork/sc-examples/tree/master/erc20-c/elrond

File: integrationTests/vm/arwen/testdata/erc20c/elrond/context.h [hunk 16]
```diff
@@ -0,0 +1,32 @@
+#include "types.h"
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply integrationTests/vm/arwen/testdata/erc20c/elrond/context.h]: We currently do this in the headers: https://github.com/ElrondNetwork/sc-examples/tree/master/erc20-c/elrond

File: integrationTests/vm/arwen/erc20_test.go (hunk 5 above)
Comment (andreibancioiu) [reply integrationTests/vm/arwen/erc20_test.go]: Only if they increase test readability. Sometimes, additional entities (such as extra constants in this case) could clutter the test.

File: integrationTests/vm/arwen/erc20_test.go (hunk 6 above)
Comment (andreibancioiu) [reply integrationTests/vm/arwen/erc20_test.go]: Only if they increase test readability. Sometimes, additional constructs (such as extra constants in this case) might clutter the test.

File: integrationTests/vm/arwen/arwenVM_test.go [hunk 17]
```diff
@@ -248,7 +248,7 @@ func deployWithTransferAndExecuteERC20(t *testing.T, numRun int, gasSchedule map
 	gasLimit := uint64(10000000000)
 	transferOnCalls := big.NewInt(5)
 
-	scCode, err := ioutil.ReadFile("./wrc20_arwen.wasm")
+	scCode, err := ioutil.ReadFile("./testdata/erc20c-v1/wrc20_arwen.wasm")
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply integrationTests/vm/arwen/arwenVM_test.go]: Only for this instance, or for all of them? We could do something like:
```
erc20/C/v1
erc20/C/v2
erc20/SOL/002
erc20/SOL/003
```

Is that ok?

[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/vm/arwen/arwenVM_test.go]: I think all the erc20 contracts can go into the same folder.

**Pull request merged**